docs/codebase-best-practices.md

```diff
@@ -0,0 +1,65 @@
+# Codebase Best Practices
+
+## General JavaScript
+
+In most cases, our [linter](how-to-setup-freecodecamp-locally?id=follow-these-steps-to-get-your-development-environment-ready) will warn of any formatting which goes against this codebase's preferred practice.
+
+## Specific TypeScript
+
+### Migrating a JavaScript File to TypeScript
+
+#### Retaining Git File History
+
+Sometimes changing the file from `<filename>.js` -> `<filename>.tsx?` causes the original file to be deleted, and a new one created, and other times the filename just changes - in terms of Git. Ideally, we want the file history to be preserved.
+
+The best bet at achieving this is to:
+
+1. Rename the file
+2. Commit with the flag `--no-verify` to prevent Husky from complaining about the lint errors
+3. Refactor to TypeScript for migration, in a separate commit
+
+> [!NOTE]
+> Editors like VSCode are still likely to show you the file has been deleted and a new one created. If you use the CLI to `git add .`, then VSCode will show the file as renamed in stage
+
+### Naming Conventions
+
+#### Types
+
+React Component Props
+
+```typescript
+type MyComponentProps = {};
+const MyComponent = (props: MyComponentProps) => {};
+```
+
+Default
+
+```typescript
+type MyObjectType = {};
+const myObject: MyObjectType = {};
```

Comment:
```suggestion
type MyObject = {};
const myObject: MyObject = {};
```
As we discussed in chat, adding the word Type to a type is one type too far.  The fact that it's a type is implicit in how it's used and TypeScript will complain if you try to use a type in a variable context.

Also, to my knowledge, no-one else does this, so we'd be unnecessarily weird.

Reply:
> no-one else does this, so we'd be unnecessarily weird.

You make this sound like a bad thing 😝 

I understand there are many (quite possibly the majority) who use the same name as the object for the type, but what about?:
```ts
type TMyObject = {};
```